### PR TITLE
[Feature] Limited Edge Dataloader (Part of auto inference).

### PR DESCRIPTION
**Pull request by yinpeiqi**
## Description

This is the limited edge dataloader. (For automatic inference helper).
For the normal node dataloader, the memory consumption of each batch is unbalance. We only fixed the number of target nodes in a batch, while the number of edges are different among different batches. Since we may encounter the cases that many high degree nodes in the same batch, and it become the bottleneck of training/inference. We develop a limited edge dataloader, which user first define two configures: `max_node` and `max_edge`. These two parameters define the maximum number of nodes and the maximum number of edges in the same batch. This dataloader mainly consider the workload of GNN inference.

The limited edge dataloader also support modify `max_node` and `max_edge` during training/inference. (The support for automatic inference).

Here we display an example of using the LimitedEdgeDataloader:
```python
import torch
import dgl
from dgl.dataloading.limited_edge import LimitedEdgeDataloader
from dgl.data import RedditDataset

if __name__ == '__main__':
    data = RedditDataset(self_loop=True)
    g = data[0].to("cuda")
    nids = torch.arange(g.number_of_nodes()).to(g.device)
    sampler = dgl.dataloading.MultiLayerFullNeighborSampler(1)
    dataloader = LimitedEdgeDataloader(
        g, nids, sampler, max_node=5000, max_edge=500000,
        shuffle=False, drop_last=False, device="cuda", num_workers=0)
    for input_nodes, output_nodes, blocks in dataloader:
        print(blocks)
```
Here I limited maximum number of nodes=5000, maximum number of edges=500000, and part of the result likes:
```
[Block(num_src_nodes=24686, num_dst_nodes=5000, num_edges=85203)]
[Block(num_src_nodes=29197, num_dst_nodes=5000, num_edges=329521)]
[Block(num_src_nodes=29056, num_dst_nodes=4775, num_edges=499702)]
[Block(num_src_nodes=43231, num_dst_nodes=5000, num_edges=477514)]
[Block(num_src_nodes=49092, num_dst_nodes=4407, num_edges=499802)]
[Block(num_src_nodes=63065, num_dst_nodes=5000, num_edges=473265)]
[Block(num_src_nodes=37184, num_dst_nodes=2961, num_edges=499686)]
[Block(num_src_nodes=72631, num_dst_nodes=4771, num_edges=498843)]
[Block(num_src_nodes=48633, num_dst_nodes=2741, num_edges=499758)]
[Block(num_src_nodes=68416, num_dst_nodes=3529, num_edges=499993)]
[Block(num_src_nodes=80186, num_dst_nodes=4128, num_edges=499989)]
[Block(num_src_nodes=78550, num_dst_nodes=3458, num_edges=499995)]
[Block(num_src_nodes=87168, num_dst_nodes=3383, num_edges=499872)]
...
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by jermainewang** [comment]
Please add some unit tests and fix the lint errors.

File: python/dgl/dataloading/limited_edge.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                         dataset,
+                         sampler,
+                         device=device,
+                         use_uva=use_uva,
+                         shuffle=shuffle,
+                         drop_last=drop_last,
+                         use_prefetch_thread=False,
+                         num_workers=num_workers)
+
+    def modify_max_edge(self, max_edge):
```
Comment (jermainewang) [inline python/dgl/dataloading/limited_edge.py:35]: Is this function used anywhere? Same for `modify_max_node` and `reset_batch_node`.

File: python/dgl/dataloading/limited_edge.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        self.dataset.max_node = max_node
+        if self.dataset.curr_iter is not None:
+            self.dataset.curr_iter.max_node = max_node
+
+    def reset_batch_node(self, node_count):
+        """Reset batch node."""
+        if self.dataset.curr_iter is not None:
+            self.dataset.curr_iter.index -= node_count
+
+    def __setattr__(self, __name, __value):
```
Comment (jermainewang) [inline python/dgl/dataloading/limited_edge.py:52]: Why need to override `__setattr__`?

**Review by jermainewang — CHANGES_REQUESTED** [review]

**Comment by jermainewang** [comment]
I don't fully understand the internal `Dataset` and `Iter` part. @BarclayII could you take a look?

File: python/dgl/dataloading/limited_edge.py (hunk 1 above)
Comment (yinpeiqi) [reply python/dgl/dataloading/limited_edge.py:35]: These three functions will be used in the auto tuner (modify the `max_node` and `max_edge` after each batch according to the memory usage).

File: python/dgl/dataloading/limited_edge.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+class LimitedEdgeDataloader(dgl.dataloading.NodeDataLoader):
+    """Limited Edge Dataloader class.
+
+    For the normal dataloader, each batch have a same batch size, which may cause the workload unbanlance
+    problem when some batches contains high degree nodes. The limited edge dataloader can support different
+    batch sizes among different batches. Users can parse two configures: max_node and max_edge, which means
+    the maximum number of nodes in a batch and the maximum number of edges in a batch. The limited edge 
+    dataloader can provide the largest batch which follows the rule.
+    """
+    def __init__(self, g, nids, sampler, max_node, max_edge, prefix_sum_in_degrees=None, \
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/dataloading/limited_edge.py]: 1. You could make `max_edge` (still think `max_degree` is better) a mandatory argument while making `max_node` an optional one. You could give it a default value (e.g., 0) meaning that the dataloader will not control minibatches by num nodes.
2. I see. Then it's ok to leave it here. Just need to update the doc.
3. I see.

File: python/dgl/dataloading/limited_edge.py (hunk 1 above)
Comment (jermainewang) [reply python/dgl/dataloading/limited_edge.py]: Is autotuner included in this PR?

File: python/dgl/dataloading/limited_edge.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yinpeiqi) [reply python/dgl/dataloading/limited_edge.py]: 1. I think set `max_node` as `None` to represent is a possible solution. But in my opinion, `max_edge` is to control the maximum number of edges inside a Graph (also known as block), It is strange to describe this item as "degree"
, since we do not say the degree of a block.
2. Okay, I will add it again. I think we can not allow `prefix_sum_in_degrees` if the `shuffle` is True.

File: python/dgl/dataloading/limited_edge.py (hunk 1 above)
Comment (yinpeiqi) [reply python/dgl/dataloading/limited_edge.py]: No, here only expose the functions that the autotuner use. I will try to add it in this PR.

File: python/dgl/dataloading/limited_edge.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/dataloading/limited_edge.py]: 1. It depends on the sampler user provides. For full neighbor sampler, you are right. But users may provide subgraph sampler, etc.
2. Then remember to have a validity check.

[the remark this answers is not in the document]
Comment (yinpeiqi) [reply python/dgl/dataloading/limited_edge.py]: Okay I changed it to `max_degree`.
